mysqld_exporter.go

```diff
@@ -264,11 +286,24 @@ func main() {
 	level.Info(logger).Log("msg", "Starting msqyld_exporter", "version", version.Info())
 	level.Info(logger).Log("msg", "Build context", version.BuildContext())
 
-	dsn = os.Getenv("DATA_SOURCE_NAME")
-	if len(dsn) == 0 {
+	if !*multiHostExporter {
```

Comment:
I have been wanting to drop the `DATA_SOURCE_NAME` mode for this exporter, as it complicates things and has a number of bugs (special chars in passwords for example).

I also would prefer to have one config parsing mode.

Reply:
Maybe we can start with this pattern, but I think I may want to refactor it in short order.

Reply:
- Tried removing the `DATA_SOURCE_NAME` env variable.  Trying to fix the tests.
- With regards to 2 modes of operation, perhaps we could run in this way with a deprecation notice about the old mode? This  might give the users some time for familiarization. And then the old mode can be removed in a future release? But i do agree that running in one mode of operation is always better.

Reply:
We should make it so that the new mode (multi-target) also supports one default case with no `target` URL param.

One thing I also think we need to do is make it so the auth part is specified by a 'module' param. This way the user/pass required for remote servers doesn't need to be repeated for each target.

This would allow for a config that only contains one default username and password to work for as many targets, as long as they all support the same password. It avoids the need to adjust the exporter config in the case of adding new targets.

Reply:
- Not sure about the support for default case with no target URL param.. Is that to handle the localhost condition?
- As far as the auth question is concerned I think that support is there in this change already. Basically we check for the target specific client in the config failing which we try the parent client (the `client` section). So if the auth is same across the fleet  one can just set up the config as follows. Perhaps it needs to be documented better in the README.

```
[client]
user = foo
password = foo123
```